### PR TITLE
Fix get_or_create_evm_token

### PR DESCRIPTION
**Pull request by nebolax**

### CODE REVIEW

File: rotkehlchen/assets/utils.py [hunk 1]
```diff
@@ -71,6 +71,14 @@ def get_or_create_evm_token(
     try:
         ethereum_token = EvmToken(identifier, form_with_incomplete_data=form_with_incomplete_data)
     except (UnknownAsset, DeserializationError):
+        try:
+            Asset(identifier).check_existence()
+            # If exists, delete the asset. Otherwise we will have a db error.
+            # Can happen if the asset is not an evm token, or if name, symbol or decimals is None
+            # while form_with_incomplete_data is False.
+            GlobalDBHandler().delete_asset_by_identifier(identifier)
```
Comment (yabirgb) [inline rotkehlchen/assets/utils.py:79]: So for me to understand what you did is delete the token if we get `UnknownAsset` but this only happens when
```
if missing_basic_data and form_with_incomplete_data is False:
```

or otherwise you are not doing any deletion since the asset was not in the database. 

In the error case the asset had all the information so it was not a problem of missing information. Also seems very dangerous to delete the asset since there could be balances in the user db depending on it and you could lead to an inconsistent state

File: rotkehlchen/assets/utils.py [hunk 2]
```diff
@@ -71,6 +71,14 @@ def get_or_create_evm_token(
     try:
         ethereum_token = EvmToken(identifier, form_with_incomplete_data=form_with_incomplete_data)
     except (UnknownAsset, DeserializationError):
+        try:
+            Asset(identifier).check_existence()
+            # If exists, delete the asset. Otherwise we will have a db error.
```
Comment (LefterisJP) [inline rotkehlchen/assets/utils.py:76]:
```suggestion
            # If it exists, delete the asset. Otherwise we will have a db error.
```

**Review by LefterisJP — COMMENTED** [review]
typo

File: rotkehlchen/assets/utils.py [hunk 3]
```diff
@@ -94,10 +111,19 @@ def get_or_create_evm_token(
...
+            GlobalDBHandler().edit_evm_token(token_data)
+        elif asset_exists is False:
+            # This can but should not raise InputError since it should not already exist
+            ethereum_token = add_ethereum_token_to_db(token_data)
+            with userdb.user_write() as cursor:
+                userdb.add_asset_identifiers(cursor, [ethereum_token.identifier])
+
+    if protocol is not None and ethereum_token.protocol != protocol:
+        object.__setattr__(ethereum_token, 'protocol', protocol)
+        GlobalDBHandler().edit_evm_token(entry=ethereum_token)
```
Comment (yabirgb) [inline rotkehlchen/assets/utils.py:126]: Why was this needed? Is an edit what is required? what about modifying the protocol in the object but not in the database?

**Review by yabirgb — COMMENTED** [review]
Not sure why the change is needed

File: rotkehlchen/assets/utils.py (hunk 3 above)
Comment (nebolax) [reply rotkehlchen/assets/utils.py:126]: Scenario that this condition addresses is:
1. In <=1.25.3 user wanted to track a curve pool token, but rotki didn't have it, so he added it manually and of course didn't specify the protocol
2. In 1.26.0 when we query curve pools, we want to set protocol to be able to properly handle price query in inquirer

Does this make sense?

Comment (yabirgb) [reply rotkehlchen/assets/utils.py:126]: okey, then seems fine. The edit won't happen always and only when required and the protocol is not set so okey for me.

File: rotkehlchen/assets/utils.py [hunk 4]
```diff
@@ -96,19 +123,27 @@ def get_or_create_evm_token(
...
+            with GlobalDBHandler().conn.critical_section():
+                ethereum_token = add_ethereum_token_to_db(
+                    token_data=token_data,
+                )
+
+            # Also update the user db inside a criticial section to avoid context switch
+            with userdb.conn.critical_section(), userdb.user_write() as cursor:
+                userdb.add_asset_identifiers(cursor, [ethereum_token.identifier])
+
+    if protocol is not None and ethereum_token.protocol != protocol:
```
Comment (LefterisJP) [inline rotkehlchen/assets/utils.py:142]: commenting here just to allow a review again.

File: rotkehlchen/assets/utils.py (hunk 3 above)
Comment (LefterisJP) [reply rotkehlchen/assets/utils.py]: I am still not sure what the PR does and why these changes.

The scenario @nebolax is describing here makes no sense.
You are modifying a very basic function used everywhere around the codebase and making it much more complicated.

You do this in order to handle a single edge case that sounds like data that should not be modified.

If a user breaks their DB that's their own fault, we can't have extra checks in every single DB related function trying to provide solutions. There is so many cases where user can do stuff to the DB they should not and we may break.

**Now for the scenario you describe** if you want to have it fixed the approach to have is a DB upgrade. Make sure that for all curve protocols (we know the addresses) the protocol is set properly during a DB upgrade.

Once that's done there there is no reason for this check anymore.

**Review by LefterisJP — COMMENTED** [review]
I rebased this PR on develop and solved conflicts.

I also fixed some stuff from Yabir's previous PR.

Now ... as I wrote in the review I am not sure what the purpose of all these changes are.

File: rotkehlchen/assets/utils.py (hunk 3 above)
Comment (nebolax) [reply rotkehlchen/assets/utils.py]: @LefterisJP the problem is that we don't know the pools during the upgrade.

File: rotkehlchen/assets/utils.py [hunk 5]
```diff
@@ -73,10 +74,21 @@ def get_or_create_evm_token(
             form_with_incomplete_data=form_with_incomplete_data,
         )
     except (UnknownAsset, DeserializationError):
+        # It can happen that the assets exists but is missing basic information. Check if it exists
```
Comment (LefterisJP) [inline rotkehlchen/assets/utils.py:77]: again when can that happen? Can someone explain this to me?

File: rotkehlchen/assets/utils.py [hunk 6]
```diff
@@ -96,19 +120,27 @@ def get_or_create_evm_token(
...
+                )
+
+            # Also update the user db inside a criticial section to avoid context switch
+            with userdb.conn.critical_section(), userdb.user_write() as cursor:
+                userdb.add_asset_identifiers(cursor, [ethereum_token.identifier])
+
+    if protocol is not None and ethereum_token.protocol != protocol:
+        # This is the case where the user added a token with the wrong protocol and later
+        # we add it with the correct one.
+        object.__setattr__(ethereum_token, 'protocol', protocol)
```
Comment (LefterisJP) [inline rotkehlchen/assets/utils.py:142]: this entire logic is still here. Why?

File: rotkehlchen/assets/utils.py [hunk 7]
```diff
@@ -96,19 +120,27 @@ def get_or_create_evm_token(
...
-        with globaldb.conn.critical_section():
-            ethereum_token = add_ethereum_token_to_db(
-                token_data=token_data,
-                globaldb=globaldb,
-            )
-
-        # Also update the user db in a criticial section to avoid a switch
-        with userdb.conn.critical_section(), userdb.user_write() as cursor:
-            userdb.add_asset_identifiers(cursor, [ethereum_token.identifier])
+        if asset_exists is True:
```
Comment (LefterisJP) [inline rotkehlchen/assets/utils.py:123]: wht is this logic? .. I am sorry I still don't get the point of this PR.

File: rotkehlchen/assets/utils.py (hunk 5 above)
Comment (yabirgb) [reply rotkehlchen/assets/utils.py:77]: One thing that can happen is that we try to recover a token that we failed to add due to a remote error or that that was not correctly filled with the user.

In the previous logic what would happen is:

1. Get the asset with `form_with_incomplete_data=False`
2. Fail since for example the decimals field is missing
3. Fetch the information
4. Try to add the token
5. Fail because the token already exists

What this PR does is:

1. If we fail to get the information check the queried properties
2. If any of the queried properties is missing fail
3. if we got the properties but any is missing update the asset
4. if the asset was not missing in the db and we got the properties add it

File: rotkehlchen/assets/utils.py [hunk 8]
```diff
@@ -96,19 +120,21 @@ def get_or_create_evm_token(
...
+            # This means that we might need to update the information in the database
+            GlobalDBHandler().edit_evm_token(token_data)
```
Comment (LefterisJP) [inline rotkehlchen/assets/utils.py:125]: Should this be "might"?

```suggestion
            # This means that we need to update the information in the database with the newly queried data
            GlobalDBHandler().edit_evm_token(token_data)
```

File: rotkehlchen/assets/utils.py [hunk 9]
```diff
@@ -96,19 +120,21 @@ def get_or_create_evm_token(
...
+        else:
+            # This can but should not raise InputError since it should not already exist.
+            # We need to wrap the whole call to add_ethereum_token_to_db since it makes
+            # the insert + when returning the EvmToken it can do a read on the database
+            with GlobalDBHandler().conn.critical_section():
+                ethereum_token = add_ethereum_token_to_db(
+                    token_data=token_data,
+                )
+
+            # Also update the user db inside a criticial section to avoid context switch
```
Comment (LefterisJP) [inline rotkehlchen/assets/utils.py:135]:
```suggestion
            # Also update the user db inside a critical section to avoid context switch
```

**Review by LefterisJP — COMMENTED** [review]
some comments.

File: rotkehlchen/assets/utils.py [hunk 10]
```diff
@@ -86,6 +98,18 @@ def get_or_create_evm_token(
             if None in (decimals, symbol, name):
                 raise NotERC20Conformant(f'Token {evm_address} is not ERC20 conformant')  # noqa: E501  # pylint: disable=raise-missing-from
 
+        if len(info) == 0 and asset_exists is True:
```
Comment (nebolax) [inline rotkehlchen/assets/utils.py:101]: I think that this check makes no sense due to:
1. `ethereum_manager.get_basic_contract_info` never returns an empty dictionary. If it doesn't find anything, it returns a dictionary with `None` values. So `len(info)` will never happen.
2. If some properties (name or symbol or decimals) are missing, then it will be caught above and `NotERC20Conformant` will be raised

Comment (LefterisJP) [reply rotkehlchen/assets/utils.py:101]: I think you are right. 

Perhaps also modify the docstring of `get_basic_contract_info()` from `Although the values might be empty.` to `Although the values might be None` to avoid any such future misunderstandings?

File: rotkehlchen/assets/utils.py [hunk 11]
```diff
@@ -73,10 +74,21 @@ def get_or_create_evm_token(
...
+            asset_exists = Asset(identifier).check_existence() is not None
+        except UnknownAsset:
+            asset_exists = False
+
         log.info(
             f'Encountered unknown asset with address '
             f'{evm_address}. Adding it to the global DB',
         )
+
+        info = {}
```
Comment (LefterisJP) [inline rotkehlchen/assets/utils.py:91]: This is also not needed
```suggestion
```

**Review by LefterisJP — APPROVED** [review]
LGTM. Small comment, but I will fix in another PR.
